File: processador/stress_cpu.py
import time
import multiprocessing

def consumir_cpu():
    """Mantém a utilização do núcleo atual próxima a 50%"""
    # Intervalo curto para amostragem rápida (0.1 segundos)
    intervalo = 0.1
    tempo_trabalho = intervalo * 0.5  # 50% de trabalho
    tempo_descanso = intervalo * 0.5  # 50% de folga

    print(f"Iniciando consumo de 50% no processo {multiprocessing.current_process().name}")
    
    while True:
        inicio = time.time()
        # Loop de processamento pesado (trabalho)
        while time.time() - inicio < tempo_trabalho:
            pass  # Executa operações o mais rápido possível
        
        # Período de ociosidade (descanso)
        time.sleep(tempo_descanso)

if __name__ == "__main__":
    # Detecta todos os núcleos disponíveis da máquina
    num_nucleos = multiprocessing.cpu_count()
    print(f"Detectados {num_nucleos} núcleos de CPU. Iniciando threads de carga...")

    processos = []
    try:
        # Cria um processo para cada núcleo para balancear a carga em 50% global
        for i in range(num_nucleos):
            p = multiprocessing.Process(target=consumir_cpu)
            p.daemon = True
            p.start()
            processos.append(p)

        # Mantém o processo principal vivo
        while True:
            time.sleep(1)
            
    except KeyboardInterrupt:
        print("\nInterrompido pelo usuário. Finalizando processos...")

File: processador/stress_cpu.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/sysinfo.h>

void consumir_cpu() {
    // Intervalo de amostragem em microssegundos (0.1 segundos)
    long intervalo = 100000; 
    long tempo_trabalho = intervalo * 0.5; // 50% trabalho
    long tempo_descanso = intervalo * 0.5; // 50% descanso

    struct timespec inicio, atual;

    while (1) {
        clock_gettime(CLOCK_MONOTONIC, &inicio);
        
        // Loop de processamento pesado
        while (1) {
            clock_gettime(CLOCK_MONOTONIC, &atual);
            long decorrido = (atual.tv_sec - inicio.tv_sec) * 1000000 + 
                             (atual.tv_nsec - inicio.tv_nsec) / 1000;
            if (decorrido >= tempo_trabalho) {
                break;
            }
        }

        // Período de descanso
        usleep(tempo_descanso);
    }
}

int main() {
    // Obtém o número de cores (núcleos) da CPU instalados
    int num_cores = sysconf(_SC_NPROCESSORS_ONLN);
    printf("Gerando carga de 50%% usando %d núcleos...\n", num_cores);

    // Cria um processo filho para cada núcleo
    for (int i = 0; i < num_cores; i++) {
        pid_t pid = fork();
        
        if (pid < 0) {
            perror("Erro ao clonar processo (fork)");
            exit(1);
        } else if (pid == 0) {
            // Processo filho entra no loop de consumo
            consumir_cpu();
            exit(0); // Garante que o filho nunca saia daqui
        }
    }

    // Processo pai apenas fica aguardando para manter a árvore viva
    while (1) {
        sleep(10);
    }

    return 0;
}